### PR TITLE
feat: Extend serviceaccount admission check for additional resource types

### LINKED ISSUE #100726: job does not report problem when pods cannot be created because referenced service account does not exist
#### What happened:
When creating a job that references a service account that doesn't exist, the job will hang indefinitely[1]. There are errors reported in the controller manager[2], but the job itself does not report anything.

[1]
```
% kubectl get job fluffy
NAME     COMPLETIONS   DURATION   AGE
fluffy   0/1                      4m32s

% kubectl get job fluffy -o jsonpath='{.status}'
{}
```

[2]
```
E0331 21:39:05.545543       1 job_controller.go:797] pods "fluffy-" is forbidden: error looking up service account default/idontexist: serviceaccount "idontexist" not found
E0331 21:39:05.545624       1 job_controller.go:402] Error syncing job: pods "fluffy-" is forbidden: error looking up service account default/idontexist: serviceaccount "idontexist" not found
I0331 21:39:05.545705       1 event.go:291] "Event occurred" object="default/fluffy" kind="Job" apiVersion="batch/v1" type="Warning" reason="FailedCreate" message="Error creating: pods \"fluffy-\" is forbidden: error looking up service account default/idontexist: serviceaccount \"idontexist\" not found"
```

#### What you expected to happen:
Something similar to pod creation -
```
% kubectl apply -f pod1
Error from server (Forbidden): error when creating "pod1": pods "bunnies" is forbidden: error looking up service account default/idontexist: serviceaccount "idontexist" not found
```

So some combination of the below would probably be fine
- The job is immediately rejected
- The job is failed
- The job status is updated

#### How to reproduce it (as minimally and precisely as possible):
example job
```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: fluffy
  namespace: default
spec:
  activeDeadlineSeconds: 10
  template:
    spec:
      containers:
      - image: busybox:latest
        imagePullPolicy: IfNotPresent
        name: box
      restartPolicy: Never
      serviceAccount: idontexist
```
pod
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: bunnies
  namespace: default
spec:
  containers:
  - image: busybox:latest
    imagePullPolicy: IfNotPresent
    name: box
  restartPolicy: Never
  serviceAccount: idontexist
```

#### Environment:
- Kubernetes version (use `kubectl version`):
```
% kubectl version
Client Version: version.Info{Major:"1", Minor:"20+", GitVersion:"v1.20.4-dirty", GitCommit:"e87da0bd6e03ec3fea7933c4b5263d151aafd07c", GitTreeState:"dirty", BuildDate:"2021-03-15T09:58:13Z", GoVersion:"go1.16.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-21T01:11:42Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
kind

### PR DESCRIPTION
**Pull request by bradbeam**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This adds support for service account lookups for batch jobs, deployments, and statefulsets.
This should prevent resources from being accepted/created if the respective service account does not exist.

#### Which issue(s) this PR fixes:

Fixes #100726

#### Does this PR introduce a user-facing change?

```release-note
ServiceAccount admission check now extends to deployments, statefulsets, and batch jobs instead of just pods.
```

### CODE REVIEW

File: plugin/pkg/admission/serviceaccount/admission.go [hunk 1]
```diff
@@ -141,35 +145,46 @@ func (s *Plugin) Admit(ctx context.Context, a admission.Attributes, o admission.
...
-	if _, isMirrorPod := pod.Annotations[api.MirrorPodAnnotationKey]; isMirrorPod {
-		return s.Validate(ctx, a, o)
+		podSpec = &o.Spec
+	case *batchapi.Job:
+		podSpec = &o.Spec.Template.Spec
+	case *appsapi.Deployment:
+		podSpec = &o.Spec.Template.Spec
+	case *appsapi.StatefulSet:
+		podSpec = &o.Spec.Template.Spec
 	}
```
Comment (RyanAoh) [inline plugin/pkg/admission/serviceaccount/admission.go:166]: I think the case of Daemonset and CronJob should also be added here.

**Comment by liggitt** [comment]
Thanks for the PR, but I don't think this direction makes sense. Resources should be able to be created in any order, and the controllers that actuate the resources wait/retry until conditions permit the desired state to be achieved.

The service account is not actually required until the pods for the workload are created. Putting synchronous checks in when creating parent resources is unnecessarily restrictive.

A better approach to resolving #100726 would be to improve the job controller so that any errors encountered when creating the pods (permission errors, validation errors, missing service account errors, quota errors, etc, etc) are surfaced in the job status better.

**Pull request closed without merging**